components/brave_wallet/browser/zcash/rust/lib.rs

```diff
@@ -411,3 +480,90 @@ impl OrchardUnauthorizedBundle {
     }
 }
 
+impl ShieldedOutput<OrchardDomain, COMPACT_NOTE_SIZE> for OrchardCompactAction {
+    fn ephemeral_key(&self) -> EphemeralKeyBytes {
+        EphemeralKeyBytes(self.ephemeral_key)
+    }
+
+    fn cmstar_bytes(&self) -> [u8; 32] {
+        self.cmx
+    }
+
+    fn enc_ciphertext(&self) -> &[u8; COMPACT_NOTE_SIZE] {
+        &self.enc_cipher_text
+    }
+}
+
+fn batch_decode(
+    fvk_bytes: &[u8; 96],
+    actions: Vec<OrchardCompactAction>
+) -> Box<BatchOrchardDecodeBundleResult> {
+    let fvk = match OrchardFVK::from_bytes(fvk_bytes) {
+        Some(fvk) => fvk,
+        None => return Box::new(BatchOrchardDecodeBundleResult::from(Err(Error::FvkError)))
+    };
+
+    let ivks = [
+        PreparedIncomingViewingKey::new(&fvk.to_ivk(OrchardScope::External)),
+        PreparedIncomingViewingKey::new(&fvk.to_ivk(OrchardScope::Internal))
+    ];
+
+    let input_actions: Result<Vec<(OrchardDomain, OrchardCompactAction)>, Error> = actions
+        .into_iter()
+        .map(|v| {
+            let nullifier_ctopt = Nullifier::from_bytes(&v.nullifier);
+            let nullifier = if nullifier_ctopt.is_none().into() {
+                Err(Error::OrchardActionFormatError)
+            } else {
+                Ok(nullifier_ctopt.unwrap())
+            }?;
+
+            let cmx_ctopt = ExtractedNoteCommitment::from_bytes(&v.cmx);
+            let cmx = if cmx_ctopt.is_none().into() {
+                Err(Error::OrchardActionFormatError)
+            } else {
+                Ok(cmx_ctopt.unwrap())
+            }?;
+
+            let ephemeral_key = EphemeralKeyBytes(v.ephemeral_key);
+            let enc_cipher_text = v.enc_cipher_text;
+
+            let compact_action = CompactAction::from_parts(nullifier, cmx, ephemeral_key, enc_cipher_text);
+            let orchard_domain = OrchardDomain::for_compact_action(&compact_action);
+
+            Ok((orchard_domain, v))
+        })
+        .collect();
+
+    let input_actions = match input_actions {
+        Ok(actions) => actions,
+        Err(e) => return Box::new(BatchOrchardDecodeBundleResult::from(Err(e.into())))
+    };
+
+    let decrypted_outputs = batch::try_compact_note_decryption(&ivks, &input_actions.as_slice())
+    .into_iter()
+    .map(|res| {
+        res.map(|((note, _recipient), _ivk_idx)| DecryptedOrchardOutput {
+            note: note
+        })
+    })
+    .filter_map(|x| x)
+    .collect::<Vec<_>>();
+
+    Box::new(BatchOrchardDecodeBundleResult::from(Ok(BatchOrchardDecodeBundleValue { outputs: decrypted_outputs })))
+}
+
+impl BatchOrchardDecodeBundle {
+    fn size(self :&BatchOrchardDecodeBundle) -> u64 {
+        return self.0.outputs.len() as u64;
+      }
+
+      fn note_value(self :&BatchOrchardDecodeBundle, index: u64) -> u32 {
+        return u32::try_from(self.0.outputs[index as usize].note.value().inner()).unwrap();
```

Comment:
`.expect("Outputs are always created from a u32, so conversion back will always succeed")` to document why this shouldn't crash outside memory corruption.

Reply:
Done